python/lsst/sims/featureScheduler/surveys/base_survey.py

```diff
@@ -0,0 +1,273 @@
+
+
+__all__ = ['BaseSurvey', 'BaseMarkovDF_survey']
+
+
+class BaseSurvey(object):
+    """A baseclass for survey objects. 
+
+    Parameters
+    ----------
+    basis_functions : list
+        List of basis_function objects
+    extra_features : list XXX--should this be a dict for clarity?
+        List of any additional features the survey may want to use
+        e.g., for computing final dither positions.
+    ignore_obs : str ('dummy')
+        If an incoming observation has this string in the note, ignore it. Handy if
+        one wants to ignore DD fields or observations requested by self. Take note,
+        if a survey is called 'mysurvey23', setting ignore_obs to 'mysurvey2' will
+        ignore it because 'mysurvey2' is a substring of 'mysurvey23'.
+    """
+    def __init__(self, basis_functions, extra_features=None,
+                 ignore_obs='dummy', survey_name='', nside=None):
+        if nside is None:
+            nside = set_default_nside()
+
+        self.nside = nside
+        self.survey_name = survey_name
+        self.nside = nside
+        self.ignore_obs = ignore_obs
+
+        self.reward = None
+        self.sequence = False  # Specifies the survey gives sequence of observations
+        self.survey_index = None
+
+        self.basis_functions = basis_functions
+
+        if extra_features is None:
+            self.extra_features = {}
+        else:
+            self.extra_features = extra_features
+        self.reward_checked = False
+
+        # Attribute to track if the reward function is up-to-date.
+        self.reward_checked = False
+
+    def add_observation(self, observation, **kwargs):
+        # ugh, I think here I have to assume observation is an array and not a dict.
+        if self.ignore_obs not in str(observation['note']):
+            for feature in self.extra_features:
+                self.extra_features[feature].add_observation(observation, **kwargs)
+            for bf in self.basis_functions:
+                bf.add_observation(observation, **kwargs)
+            self.reward_checked = False
+
+    def _check_feasibility(self, conditions):
+        """
+        Check if the survey is feasable in the current conditions
+        """
+        for bf in self.basis_functions:
+            result = bf.check_feasibility(conditions)
+            if not result:
+                return result
+        return result
+
+    def calc_reward_function(self, conditions):
+        """
+        Parameters
+        ----------
+        conditions : lsst.sims.featureScheduler.features.Conditions object
+
+        Returns
+        -------
+        reward : float (or array)
+
+        """
+        if self._check_feasability():
+            self.reward = 0
+        else:
+            # If we don't pass feasability
+            self.reward = -np.inf
+
+        self.reward_checked = True
+        return self.reward
+
+    def __call__(self, conditions):
```

Comment:
Rename method 'request observation' ?

Reply:
It seems like this isn't called directly anymore, but that the reward functions are and then balanced at a higher level?

Reply:
Let's go with "generate_observations". It gets called in `core_scheduler`. `_fill_queue` loops over all the surveys to find the one with the highest reward, then goes and has that one generate a list of observations.